src/queue.jl

```diff
@@ -1,64 +1,156 @@
 # Utilities for working with HSA queues
 
-"""
-An HSA queue.
-Each queue is uniquely associated with an agent.
-"""
-mutable struct HSAQueue
-    agent::HSAAgent
-    queue::Ref{Ptr{HSA.Queue}}
-    active::Bool
+@static if VERSION >= v"1.7-"
+    """
+    An HSA queue.
+    Each queue is uniquely associated with an agent.
+    """
+    mutable struct HSAQueue
+        agent::HSAAgent
+        queue::Ptr{HSA.Queue}
+        status::HSA.Status
+        cond::Base.AsyncCondition
+        @atomic active::Bool
+    end
+else
+    """
+    An HSA queue.
+    Each queue is uniquely associated with an agent.
+    """
+    mutable struct HSAQueue
+        agent::HSAAgent
+        queue::Ptr{HSA.Queue}
+        status::HSA.Status
+        cond::Base.AsyncCondition
+        active::Bool
+    end
+    @inline function replacefield!(value, name::Symbol, expected, desired)
+        y = getfield(value, name)
+        ok = y === expected
+        if ok
+            setfield!(value, name, desired)
+        end
+        return (; old=y, success=ok)
+    end
 end
-
+const QUEUES = Dict{Ptr{HSA.Queue},WeakRef}()
 const DEFAULT_QUEUES = IdDict{HSAAgent,HSAQueue}()
 
+async_send(data::Ptr{Cvoid}) = ccall(:uv_async_send, Cint, (Ptr{Cvoid},), data)
+
+function queue_error_handler(status::HSA.Status, _queue::Ptr{HSA.Queue}, queue_obj_ptr::Ptr{Cvoid})::Nothing
+    if status != HSA.STATUS_SUCCESS
+        queue = unsafe_pointer_to_objref(queue_obj_ptr)::HSAQueue
```

Comment:
I think there is the potential risk here that the `queue` got GC'd.

Reply:
If the queue gets GC'd, the finalizer will first destroy the queue. This should prevent async callbacks from firing after the finalizer completes, since destroying the queue waits for any pending callbacks to complete.